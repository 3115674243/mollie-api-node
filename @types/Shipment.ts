declare namespace Mollie {
  interface Shipment {
    lines: Array<OrderLine>;
    tracking?: Tracking;

    // Access token parameters
    testmode?: boolean;
  }

  interface OrderShipment extends Shipment {
    orderId?: string,
  }

  interface Tracking {
    carrier: string;
    code: string;
    url?: string;
  }
}
