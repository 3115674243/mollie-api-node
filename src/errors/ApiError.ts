import { MollieApiError, MollieApiErrorLinks, Url } from '../data/global';
import { cloneDeep, get, has } from 'lodash';

/**
 * @since 3.0.0
 */
export default class ApiError extends Error {
  public constructor(message: string, protected title?: string, protected status?: number, protected field?: string, protected links?: MollieApiErrorLinks) {
    super(message);
    this.name = 'ApiError';
  }

  /**
   * Get the error message
   *
   * @returns The error message
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getMessage(): string {
    return this.message;
  }

  /**
   * Get the field name that contains an error
   *
   * @returns The error field
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getField(): string | undefined {
    return this.field;
  }

  /**
   * Get the API status code
   *
   * @returns The status code
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getStatusCode(): number | undefined {
    return this.status;
  }

  /**
   * Get the documentation URL
   *
   * @returns The documentation URL
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getDocumentationUrl(): string {
    return this.getUrl('documentation');
  }

  /**
   * Get the dashboard URL
   *
   * @returns The dashboard URL
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getDashboardUrl(): string {
    return this.getUrl('dashboard');
  }

  /**
   * Check if the link exists
   *
   * @param key - Link name
   *
   * @returns Whether the link exists
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public hasLink(key: string): boolean {
    return has(this.links, key);
  }

  /***
   * Retrieve a link by name
   *
   * @param key - The link name
   *
   * @returns A link to the resource
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getLink(key: string): Url {
    return get(this.links, key);
  }

  /**
   *
   * @param {string} key
   *
   * @returns {string}
   *
   * @since 3.0.0
   *
   * @public ✓ This method is part of the public API
   */
  public getUrl(key: string): string {
    return get(this.getLink(key), 'href');
  }

  public toString(): string {
    return `ApiError: ${this.message}`;
  }

  /**
   * Create an `ApiError` from a raw error format
   *
   * @param error - A raw Mollie API error
   *
   * @returns A new `ApiError`
   *
   * @since 3.0.0
   */
  public static createFromResponse(error: MollieApiError): ApiError {
    return new ApiError(get(error, 'data.detail'), get(error, 'data.title'), get(error, 'data.status'), get(error, 'data.field'), cloneDeep(get(error, 'data._links')));
  }
}
