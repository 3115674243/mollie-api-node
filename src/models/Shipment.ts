import Model from '../model';
import { IShipment } from '../types/shipment';

/**
 * The `Shipment` model
 *
 * {@link IShipment}
 */
export default class Shipment extends Model implements IShipment {
  public static resourcePrefix = 'shp_';

  public resource = 'shipment';
  public id = null;
  public orderId = null;
  public createdAt = null;
  public tracking = null;
  public lines = null;
  public _links = {
    self: null,
    order: null,
    documentation: null,
  };

  // Access token parameters
  public testmode?: boolean;

  /**
   * Shipment constructor
   *
   * @public âœ“ This constructor is part of the public API
   */
  public constructor(props?: Partial<IShipment>) {
    super();

    Object.assign(this, props);
  }
}
