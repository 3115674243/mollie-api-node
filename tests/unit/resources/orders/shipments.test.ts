import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import OrdersShipmentsResource from '@resources/orders/shipments';
import Shipment from '@models/Shipment';
import response from '@tests/unit/__stubs__/shipments.json';
import ApiError from '@errors/ApiError';

const mock = new MockAdapter(axios);

const props = {
  id: 'shp_3wmsgCJN4U',
  orderId: 'ord_kEn1PlbGa',
  lines: [],
  tracking: {
    carrier: 'PostNL',
    code: '3SKABA000000000',
    url: 'http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C',
  },
};

describe('orders_shipments', () => {
  let ordersShipments: OrdersShipmentsResource;
  beforeEach(() => {
    ordersShipments = new OrdersShipmentsResource(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(OrdersShipmentsResource.resource).toBe('orders_shipments');
    expect(OrdersShipmentsResource.model).toBe(Shipment);
  });

  describe('.create()', () => {
    mock.onPost(`/orders/${props.orderId}/shipments`).reply(200, response._embedded.shipments[0]);

    it('should return a shipment instance', () =>
      ordersShipments.create(props).then(result => {
        expect(result).toBeInstanceOf(Shipment);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      ordersShipments.create(props, (err, result) => {
        expect(result).toBeInstanceOf(Shipment);
        expect(result).toMatchSnapshot();
        done();
      });
    });
  });

  describe('.get()', () => {
    const error = { detail: 'The orders_shipments id is invalid' };

    mock.onGet(`/orders/${props.orderId}/shipments/${props.id}`).reply(200, response._embedded.shipments[0]);
    mock.onGet(`/orders/${props.orderId}/shipments/foo`).reply(500, error);

    it('should return a shipment instance', () =>
      ordersShipments.get(props.id, { orderId: props.orderId }).then(result => {
        expect(result).toBeInstanceOf(Shipment);
        expect(result).toMatchSnapshot();
      }));

    it('should return an error for non-existing IDs', done =>
      ordersShipments
        .get('foo', { orderId: props.orderId })
        .then(result => expect(result).toBeUndefined())
        .catch(error => {
          expect(error).toBeInstanceOf(ApiError);
          done();
        }));
  });

  describe('.list()', () => {
    const error = { detail: 'The order id is invalid' };
    mock.onGet(`/orders/${props.orderId}/shipments`).reply(200, response);

    it('should return a list of all payment refunds', () =>
      ordersShipments.list({ orderId: props.orderId }).then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should throw an error if "paymentId" is not set', done => {
      ordersShipments
        .list(undefined)
        .then(result => expect(result).toBeUndefined())
        .catch(err => {
          expect(err).toBeInstanceOf(ApiError);
          expect(err.getMessage()).toEqual(error.detail);
          done();
        });
    });

    it('should work with a callback', done => {
      ordersShipments
        .withParent({
          resource: 'order',
          id: props.orderId,
        })
        .list((err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        });
    });

    it('should work with a callback', done => {
      ordersShipments
        .withParent({
          resource: 'order',
          id: props.orderId,
        })
        .list((err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        });
    });

    it('should work with a Promise and with .withParent()', done => {
      ordersShipments
        .withParent({
          resource: 'order',
          id: props.orderId,
        })
        .list(undefined)
        .then(result => {
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        })
        .catch(error => expect(error).toBeUndefined());
    });
  });
});
