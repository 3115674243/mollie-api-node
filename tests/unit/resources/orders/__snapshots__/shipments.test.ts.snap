// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orders_shipments .create() should return a shipment instance 1`] = `
Shipment {
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
      "type": "text/html",
    },
    "order": Object {
      "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
      "type": "application/hal+json",
    },
  },
  "createdAt": "2018-08-09T14:33:54+00:00",
  "id": "shp_3wmsgCJN4U",
  "lines": Array [
    Object {
      "_links": Object {
        "imageUrl": Object {
          "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
          "type": "text/html",
        },
        "productUrl": Object {
          "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
          "type": "text/html",
        },
      },
      "createdAt": "2018-08-02T09:29:56+00:00",
      "discountAmount": Object {
        "currency": "EUR",
        "value": "100.00",
      },
      "id": "odl_dgtxyl",
      "isCancelable": true,
      "name": "LEGO 42083 Bugatti Chiron",
      "orderId": "ord_pbjz8x",
      "quantity": 1,
      "resource": "orderline",
      "sku": "5702016116977",
      "status": "shipping",
      "totalAmount": Object {
        "currency": "EUR",
        "value": "299.00",
      },
      "type": "physical",
      "unitPrice": Object {
        "currency": "EUR",
        "value": "399.00",
      },
      "vatAmount": Object {
        "currency": "EUR",
        "value": "51.89",
      },
      "vatRate": "21.00",
    },
  ],
  "mode": undefined,
  "orderId": "ord_kEn1PlbGa",
  "resource": "shipment",
  "testmode": undefined,
  "tracking": Object {
    "carrier": "PostNL",
    "code": "3SKABA000000000",
    "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
  },
}
`;

exports[`orders_shipments .create() should work with a callback 1`] = `
Shipment {
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
      "type": "text/html",
    },
    "order": Object {
      "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
      "type": "application/hal+json",
    },
  },
  "createdAt": "2018-08-09T14:33:54+00:00",
  "id": "shp_3wmsgCJN4U",
  "lines": Array [
    Object {
      "_links": Object {
        "imageUrl": Object {
          "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
          "type": "text/html",
        },
        "productUrl": Object {
          "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
          "type": "text/html",
        },
      },
      "createdAt": "2018-08-02T09:29:56+00:00",
      "discountAmount": Object {
        "currency": "EUR",
        "value": "100.00",
      },
      "id": "odl_dgtxyl",
      "isCancelable": true,
      "name": "LEGO 42083 Bugatti Chiron",
      "orderId": "ord_pbjz8x",
      "quantity": 1,
      "resource": "orderline",
      "sku": "5702016116977",
      "status": "shipping",
      "totalAmount": Object {
        "currency": "EUR",
        "value": "299.00",
      },
      "type": "physical",
      "unitPrice": Object {
        "currency": "EUR",
        "value": "399.00",
      },
      "vatAmount": Object {
        "currency": "EUR",
        "value": "51.89",
      },
      "vatRate": "21.00",
    },
  ],
  "mode": undefined,
  "orderId": "ord_kEn1PlbGa",
  "resource": "shipment",
  "testmode": undefined,
  "tracking": Object {
    "carrier": "PostNL",
    "code": "3SKABA000000000",
    "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
  },
}
`;

exports[`orders_shipments .get() should return a shipment instance 1`] = `
Shipment {
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
      "type": "text/html",
    },
    "order": Object {
      "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
      "type": "application/hal+json",
    },
  },
  "createdAt": "2018-08-09T14:33:54+00:00",
  "id": "shp_3wmsgCJN4U",
  "lines": Array [
    Object {
      "_links": Object {
        "imageUrl": Object {
          "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
          "type": "text/html",
        },
        "productUrl": Object {
          "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
          "type": "text/html",
        },
      },
      "createdAt": "2018-08-02T09:29:56+00:00",
      "discountAmount": Object {
        "currency": "EUR",
        "value": "100.00",
      },
      "id": "odl_dgtxyl",
      "isCancelable": true,
      "name": "LEGO 42083 Bugatti Chiron",
      "orderId": "ord_pbjz8x",
      "quantity": 1,
      "resource": "orderline",
      "sku": "5702016116977",
      "status": "shipping",
      "totalAmount": Object {
        "currency": "EUR",
        "value": "299.00",
      },
      "type": "physical",
      "unitPrice": Object {
        "currency": "EUR",
        "value": "399.00",
      },
      "vatAmount": Object {
        "currency": "EUR",
        "value": "51.89",
      },
      "vatRate": "21.00",
    },
  ],
  "mode": undefined,
  "orderId": "ord_kEn1PlbGa",
  "resource": "shipment",
  "testmode": undefined,
  "tracking": Object {
    "carrier": "PostNL",
    "code": "3SKABA000000000",
    "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
  },
}
`;

exports[`orders_shipments .list() should return a list of all payment refunds 1`] = `
List [
  Shipment {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-08-09T14:33:54+00:00",
    "id": "shp_3wmsgCJN4U",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "isCancelable": true,
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_pbjz8x",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "shipping",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "mode": undefined,
    "orderId": "ord_kEn1PlbGa",
    "resource": "shipment",
    "testmode": undefined,
    "tracking": Object {
      "carrier": "PostNL",
      "code": "3SKABA000000000",
      "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
    },
  },
]
`;

exports[`orders_shipments .list() should work with a Promise and with .withParent() 1`] = `
List [
  Shipment {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-08-09T14:33:54+00:00",
    "id": "shp_3wmsgCJN4U",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "isCancelable": true,
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_pbjz8x",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "shipping",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "mode": undefined,
    "orderId": "ord_kEn1PlbGa",
    "resource": "shipment",
    "testmode": undefined,
    "tracking": Object {
      "carrier": "PostNL",
      "code": "3SKABA000000000",
      "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
    },
  },
]
`;

exports[`orders_shipments .list() should work with a callback 1`] = `
List [
  Shipment {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-08-09T14:33:54+00:00",
    "id": "shp_3wmsgCJN4U",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "isCancelable": true,
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_pbjz8x",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "shipping",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "mode": undefined,
    "orderId": "ord_kEn1PlbGa",
    "resource": "shipment",
    "testmode": undefined,
    "tracking": Object {
      "carrier": "PostNL",
      "code": "3SKABA000000000",
      "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
    },
  },
]
`;

exports[`orders_shipments .list() should work with a callback 2`] = `
List [
  Shipment {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/shipments-api/get-shipment",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_kEn1PlbGa",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/order/ord_kEn1PlbGa/shipments/shp_3wmsgCJN4U",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-08-09T14:33:54+00:00",
    "id": "shp_3wmsgCJN4U",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "isCancelable": true,
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_pbjz8x",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "shipping",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "mode": undefined,
    "orderId": "ord_kEn1PlbGa",
    "resource": "shipment",
    "testmode": undefined,
    "tracking": Object {
      "carrier": "PostNL",
      "code": "3SKABA000000000",
      "url": "http://postnl.nl/tracktrace/?B=3SKABA000000000&P=1016EE&D=NL&T=C",
    },
  },
]
`;
